File: clients/FSharp/Model/RepairProperties.fs
#nowarn "0058"
namespace Aicup2020.Model
type RepairProperties = {
    ValidTargets: EntityType[];
    Power: int;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.ValidTargets.Length
        this.ValidTargets |> Array.iter (fun value ->
            writer.Write (int value)
        )
        writer.Write this.Power
    static member readFrom(reader: System.IO.BinaryReader) = {
        ValidTargets = [|for _ in 1 .. reader.ReadInt32() do
            yield reader.ReadInt32() |> enum
        |]
        Power = reader.ReadInt32()
    }

File: clients/Go/model/repair_properties.go
package model

import "io"
import . "aicup2020/stream"

type RepairProperties struct {
    ValidTargets []EntityType
    Power int32
}
func NewRepairProperties(validTargets []EntityType, power int32) RepairProperties {
    return RepairProperties {
        ValidTargets: validTargets,
        Power: power,
    }
}
func ReadRepairProperties(reader io.Reader) RepairProperties {
    result := RepairProperties {}
    result.ValidTargets = make([]EntityType, ReadInt32(reader))
    for i := range result.ValidTargets {
        result.ValidTargets[i] = ReadEntityType(reader)
    }
    result.Power = ReadInt32(reader)
    return result
}
func (value RepairProperties) Write(writer io.Writer) {
    WriteInt32(writer, int32(len(value.ValidTargets)))
    for _, ValidTargetsElement := range value.ValidTargets {
        WriteInt32(writer, int32(ValidTargetsElement))
    }
    WriteInt32(writer, value.Power)
}
